lib/src/rules/avoid_unstable_final_fields.dart

```diff
@@ -0,0 +1,685 @@
+non-final field, or it could be an explicitly declared getter with a body
+that isn't known to return the same value each time it is called.
+
+The underlying motivation for this rule is that if it is followed then a final
+field is an immutable property of an object. This is important for correctness
+because it is then safe to assume that the value does not change during the
+execution of an algorithm. In contrast, it may be necessary to re-check any
+other getter repeatedly if it is not known to have this property. Similarly,
+it is safe to cache the immutable property in a local variable and promote it,
+but for any other property it is necessary to check repeatedly that the
+underlying property hasn't changed since it was promoted.
+
+**BAD:**
+```dart
+class A {
+  final int i;
+  A(this.i);
+}
+
+var j = 0;
+
+class B1 extends A {
+  int get i => ++j + super.i; // LINT.
+  B1(super.i);
+}
+
+class B2 implements A {
+  int i; // LINT.
+  B2(this.i);
+}
+```
+
+**GOOD:**
+```dart
+class C {
+  final int i;
+  C(this.i);
+}
+
+class D1 implements C {
+  late final int i = someExpression; // OK.
+}
+
+class D2 extends C {
+  int get i => super.i + 1; // OK.
+  D2(super.i);
+}
+```
+
+''';
+
+class AvoidUnstableFinalFields extends LintRule {
+  AvoidUnstableFinalFields()
+      : super(
+            name: 'avoid_unstable_final_fields',
+            description: _desc,
+            details: _details,
+            group: Group.errors);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this, context);
+    registry.addFieldDeclaration(this, visitor);
+    registry.addMethodDeclaration(this, visitor);
+  }
+}
+
+bool _isLocallyStable(Element element) {
+  if (element is PropertyAccessorElement &&
+      element.isSynthetic &&
+      element.correspondingSetter == null) {
+    // This is a final, non-local variable, and they are stable.
+    var metadata = element.variable.metadata;
+    if (metadata.isNotEmpty) {
+      for (var elementAnnotation in metadata) {
+        var metadataElement = elementAnnotation.element;
+        if (metadataElement is ConstructorElement) {
+          var metadataOwner = metadataElement.declaration.enclosingElement;
+          if (metadataOwner.isDartCoreObject) {
+            // A declaration with `@Object()` is not considered stable.
+            return false;
+          }
+        }
+      }
+    }
+    return true;
+  } else if (element is FunctionElement) {
+    // A tear-off of a top-level function or static method is stable,
+    // local functions and function literals are not.
+    return element.isStatic;
+  } else if (element is ClassElement) {
+    // A reified type of a class is stable.
+    return true;
+  } else if (element is MethodElement) {
+    // An instance method tear-off is never stable.
+    return false;
+  }
+  // TODO(eernst): Any cases still missing?
```

Comment:
The latest version of the analyze package (a PR to move to it is in flight as I type) uses the new classes `EnumElement` and `MixinElement` rather than using `ClassElement` for all of the cases. You'll probably want to add those cases the next time you rebase your branch.